### PR TITLE
Add ShapeElementUserObject base class

### LINKED ISSUE #5913: Need the ability for UserObjects to contribute to the Jacobian
* Add flag to ```Assembly``` that is set to true as soon as even one UO requires the shapes (we have a setter that can only set to true - there is no reason to ever unset this).
* Query that flag in ```ComputeUserObjectsThread``` (we'll retrieve a const ref from assembly)
* call FEProblem::prepareShapes in  ```ComputeUserObjectsThread::onElement``` if the flag is true
* The new base class ```ShapeElementUserObject``` will provide references to _phi and _grad_phi 

Ping @SudiptaBiswas

### PR DESCRIPTION
**Pull request by dschwen**
Derived from `ElementUserObject` this class makes `_phi` and `_grad_phi` available and correctly initialized whenever `_fe_problem.currentlyComputingJacobian()` returns true.

Closes #5913

### CODE REVIEW

**Comment by dschwen** [comment]
I should probably add a test class in moose-test for this...

**Comment by SudiptaBiswas** [comment]
As I modified the ComputeGrainForceAndTorque class, deriving it from ShapeElementUserObject, instead of ElementUserObject, it's throwing segmentation fault from the following line in ComputeUserObjectsThread::onElement,
  _fe_problem.prepareShapes(_user_object_shape_variables[i], _tid);

**Comment by dschwen** [comment]
problem is already in ```ComputeUserObjectsThread.C:31```. ```_tid``` is not set at that point.

**Comment by friedmud** [comment]
I will try to review this today - but it is a crazy day. Give me a bit of
time.

File: framework/src/userobject/ShapeElementUserObject.C [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+template<>
+InputParameters validParams<ShapeElementUserObject>()
+{
+  InputParameters params = validParams<ElementUserObject>();
+  return params;
+}
+
+ShapeElementUserObject::ShapeElementUserObject(const InputParameters & parameters) :
+    ElementUserObject(parameters),
+    _phi(_assembly.phi()),
```
Comment (friedmud) [inline framework/src/userobject/ShapeElementUserObject.C]: This won't work... And demonstrates what I was worried about.

The issue is; what if this UserObject is coupled to multiple variables?  If one of those variables is linear Lagrange and the other is cubic hermite: then phi will be wrong.

To do this properly you're going to need to have an interface similar to Kernels and everything else (which is the right thing to do anyway) and have a special `computeElementUserObjectJacobianThread` that does the right thing.

I haven't yet thought of _how_ to store these Jacobians... I'll give it some more thought.

Comment (dschwen) [inline framework/src/userobject/ShapeElementUserObject.C]: This is the same thing that KernelBase does, and in Kernels we also utilize the _phis as derivatives of the coupled variables in computeOffDiagJacobian. What is Kernel doing differently so that it is not an issue there?

Comment (friedmud) [inline framework/src/userobject/ShapeElementUserObject.C]: The difference is that Kernels et. al. have a dedicated `computeJacobian()` that can be called for each coupled variable.  In between each call the shapes are swapped out behind the scenes to make sure that `phi`, etc. are correct for the derivative MOOSE is currently asking for.

See here:

https://github.com/idaholab/moose/blob/devel/framework/src/base/ComputeFullJacobianThread.C#L63

See how the shapes are re-prepared (why are they always preparing? :-) before every invocation of `Kernel::computeJacobian()`?  That makes sure that for each call the `phi` variables contain the correct shapes.

Basically: the right thing to do here is add a `computeJacobian()` call to UserObjects.  That way the Jacobian can be computed for each coupled variable.

Generalizing the Jacobian value storage will be difficult though... still thinking about that.

**Comment by dschwen** [comment]
Btw. this fails ```materials/stateful_prop.test_older_mpi_threads``` which is passed on my machine (in opt and dbg). I don't think that failure is related to this PR.

File: framework/src/userobject/ShapeElementUserObject.C (hunk 1 above)
Comment (dschwen) [inline framework/src/userobject/ShapeElementUserObject.C]: Ok, got it. Of course the ```_j``` indexes only the local dofs of _one_ variable, not all local DOFs on the element. In ```ComputeUserObjectsThread``` I loop over all variables and prepare the phis, but what that does is just overwrite them a bunch of times.

We'd need an ```executeJacobian(unsigne int jvar)``` method that gets called after a variable is prepared. I still maintain the point that Jacobian storage should be up to the user (just like the interface for the ```residual``` data of teh UserObject).

**Comment by dschwen** [comment]
Question is how to utilize the ```_fe_problem.couplingEntries(_tid)``` data. It would probably have to loop through all variables coupled by the UO as we have no way of knowing how the UO jacobian contribution would end up getting used.

**Comment by dschwen** [comment]
eb9027a is an idea how to salvage this PR, just contribution to the discussion. It...

* implements an ```executeJacobian(unsigned int)``` method during which the shape functions are correctly initialized
* implements a way to interrogate an ```ElementUserObject``` if the ```executeJacobian``` should be called for a given variable

The ```ShapeElementUserObject``` registers all its coupled variables. There are two registration tiers here:

1. A global registration in assembly, i.e. a list of non-linear variable numbers for which shape functions need to be prepared. As soon as even one UO registers a coupled variable for Jacobian calculation purposes it is added to that global list
2. A local flag that determines if ```executeJacobian``` will be called for a given UO. We could drop this and opt to always call executeJacobian and rely in the implementation of executeJacobian to bail out immediately if nothing needs to be calculated for the current variable.

**Comment by friedmud** [comment]
I haven't had time to review everything yet... but I would rather the function be called `computeOffDiagJacobian()` and function just like it does everywhere else in MOOSE.  There should be just one way to do this.

File: framework/include/userobject/ElementUserObject.h [hunk 2]
```diff
@@ -63,12 +63,20 @@ class ElementUserObject :
...
+  virtual void executeJacobian(unsigned int /*jvar*/) {}
+
+  /**
    * Must override.
    *
    * @param uo The UserObject to be joined into _this_ object.  Take the data from the uo object and "add" it into the data for this object.
    */
   virtual void threadJoin(const UserObject & uo) = 0;
 
+  bool requestedJacobian(unsigned int var) const;
```
Comment (friedmud) [inline framework/include/userobject/ElementUserObject.h]: Make this return a vector of variables instead... that way we can combine and loop over them on the outside.  We can do some preprocessing in `ThreadedElementLoop::subdomainChanged()`.

If we get this right we may think about extending this to Kernels, etc!

File: framework/src/base/Assembly.C [hunk 3]
```diff
@@ -1522,6 +1524,19 @@ Assembly::addCachedJacobianContributions(SparseMatrix<Number> & jacobian)
...
+Assembly::registerUserObjectShapeVariable(unsigned int var)
+{
+  std::vector<unsigned int> & reg = _user_object_shape_variables;
+
+  // add variable number to the list of registered variables
+  reg.push_back(var);
+
+  // sort and keep elements unique
+  std::sort(reg.begin(), reg.end());
+  reg.erase(std::unique(reg.begin(), reg.end()), reg.end());
```
Comment (friedmud) [inline framework/src/base/Assembly.C]: Why not just use a `set`?

File: framework/src/userobject/ElementUserObject.C [hunk 4]
```diff
@@ -52,3 +52,8 @@ ElementUserObject::ElementUserObject(const InputParameters & parameters) :
     addMooseVariableDependency(coupled_vars[i]);
 }
 
+bool
+ElementUserObject::requestedJacobian(unsigned int var) const
+{
+  return var < _requested_jacobian_flag.size() && _requested_jacobian_flag[var];
```
Comment (friedmud) [inline framework/src/userobject/ElementUserObject.C]: Ummm... this is a bad idea.  Just use a `set`.  With thousands of variables this could be unnecessarily large... with just a few variables `set` is going to be fast anyway.

File: framework/src/base/Assembly.C (hunk 3 above)
Comment (dschwen) [inline framework/src/base/Assembly.C]: Because it is written to only at the start of the simulation but iterated over a LOT. A vector send to be the more appropriate days structure.

**Comment by friedmud** [comment]
Let's remove the intermediate base class (`ShapeElementUserObject`).  Just add the ability to request Jacobian calculations for any UserObject and do it the same way everywhere.

No one wants to remember to request Jacobians except not sometimes when you happen to have that intermediate base class in your tree.  Uniformity wins over convenience.

File: framework/src/userobject/ElementUserObject.C (hunk 4 above)
Comment (dschwen) [inline framework/src/userobject/ElementUserObject.C]: How is this ever large? Even with thousands of variables it is at most a few kB!

**Comment by dschwen** [comment]
I'm not a fan of calling this computeOffDiagJacobian. First of all we have no computer residual here, secondly the concept of on or off diagonal makes no sense, we will not have a return value, and we are not really computing a Jacobian anyways, just terms that enter a Jacobian.

**Comment by friedmud** [comment]
I can see it in MOOSE training now:

"When you're using Kernel, BoundaryCondition, DiracKernel, DGKernel, ScalarKernel override `computeOffDiagJacobian()`.

However, make sure you remember the crazy oddball system: when you want to compute Jacobians for UserObjects the function is called `executeJacobian()`.  I can't even remember why it's different from the others... and it's often the case that I forget this myself and I can't figure out why my Jacobian values aren't getting computed for my UserObjects (because calling it `computeOffDiagJacobian()` doesn't generate an error of any kind... it just doesn't get called).

In fact: I don't even understand why we want to "execute" Jacobians at all!  Jacobians don't *do* anything... so execute doesn't even make sense!

Oh well... it's just history now.... just make sure you remember the one screwy system in MOOSE that has to be different!"

....

No: It should match everything else.

As for On/Off diagonal... two things:

1.  UserObjects that inherit from one of the `Variable` classes (like `ElementVariablePostprocessor`) _should_ actually have the opportunity to override `computeJacobian()` that does the on-diagonal computation.  This is because those classes take a `variable` parameter that is _the_ variable they are working on.

2.  UserObjects that only couple to variables (by calling `coupledValue()`, etc.) should only have `computeOffDiagJacobian()`... because that's exactly how every other system in MOOSE works: if you _couple_ to it then you should expect to be able to compute a Jacobian value for it in `computeOffDiagJacobian()`.

File: framework/src/userobject/ElementUserObject.C (hunk 4 above)
Comment (friedmud) [inline framework/src/userobject/ElementUserObject.C]: With n_threads copies of UserObjects and lots of them...

File: framework/src/base/Assembly.C (hunk 3 above)
Comment (friedmud) [inline framework/src/base/Assembly.C]: Iterating over a `set` with an iterator is going to be WAY in the noise.  When you compare an iterator movement and dereference to incrementing a local counter and then indexing into the vector (like you are doing now) there's no way there's any difference that would be measureable in MOOSE.

A `set` shows everyone exactly what this is: a unique list of IDs.  No one can ever screw it up by "side loading" values into it... and it's self documenting.

**Comment by dschwen** [comment]
So how is the signature for ```computeOffDiagJacobian()``` supposed to look like? Is it returning a ```Real``` value like the call in Kernel? What is that value supposed to be? There is a slim chance of this working in Postprocessors. And even there we'd have to _massively_ change the interface. So for consistency's sake we'll have a ```computeResidual``` in Postprocessors now (we wouldn't want Postprocessors to be a _crazy oddball system_, right)?! 
There is **no** chance of this being feasible for UserObjects. Take the object @SudiptaBiswas is working on (and that is a simple example). It computes two std::vectors of ```RealGradient```s. How is that ```computeOffDiagJacobian()``` method going to look here?

File: framework/src/base/Assembly.C (hunk 3 above)
Comment (dschwen) [inline framework/src/base/Assembly.C]: Ok

**Comment by friedmud** [comment]
I never said it should return a Real.  I said it should be called `computeOffDiagJacobian()`.  If you accidentally define it as returning a Real you will get an error because it will be shadowing a virtual function.

File: framework/include/userobject/ElementUserObject.h (hunk 2 above)
Comment (dschwen) [inline framework/include/userobject/ElementUserObject.h]: Wait, we already loop over a vector of variables, the ones registered in assembly! We initialize the shape function once for every user object, in the loop over the UOs we check if we really need to call ```executeJacobian```. I think we can just drop this alltogether and call an empty function (or bail out depending on the value of jvar).

File: framework/src/userobject/ElementUserObject.C (hunk 4 above)
Comment (dschwen) [inline framework/src/userobject/ElementUserObject.C]: Yeah, if it is a set it is for sure cheaper to _just call_ ```executeJacobian```. I think I'd rather remove this mechanism entirely (I've already written a comment to that effect a few hours ago).

File: framework/src/userobject/ShapeElementUserObject.C (hunk 1 above)
Comment (dschwen) [inline framework/src/userobject/ShapeElementUserObject.C]: Ok, this comment is effectively outdated now, right?

**Comment by friedmud** [comment]
Also: we don't need `registerUserObjectShapeVariable()`.

In `subdomainChanged()` in your ThreadedElement loop... just loop over the active UserObjects and merge their _sets_ of variable IDs to create the list of variables you need Jacobians for.  Then loop over that.

As for how to detect that a UserObject _wants_ to compute a Jacobian... we might be able to directly key off of whether or not `coupled()` has been called for a variable.  Or, we can do something more explicit.

File: framework/include/userobject/ElementUserObject.h (hunk 2 above)
Comment (friedmud) [inline framework/include/userobject/ElementUserObject.h]: Yes... with the caveat that we don't even need a list in Assembly (see my comment back over on the PR).

**Comment by dschwen** [comment]
I added ```registerUserObjectShapeVariable``` because I did not want to initialize tons of shape functions and loop over all user objects, if they will be unused because barely anybody is computing Jacobian terms.

File: framework/include/userobject/ElementUserObject.h (hunk 2 above)
Comment (friedmud) [inline framework/include/userobject/ElementUserObject.h]: Wait - I got mixed up here... _No_... we should keep `requestedJacobian()` (but make it a set) and jettison the registration in Assembly.  That's what I meant ;-)

File: framework/src/userobject/ElementUserObject.C (hunk 4 above)
Comment (friedmud) [inline framework/src/userobject/ElementUserObject.C]: I do think the in loop checking of each one is unnecessary... but it may still be a decent plan to form a list of only the variables that you need UserObject Jacobians for.

There could be quite a LOT of time spent calling `prepareShapes()`, etc. for all variables in the system when we really don't need to.

File: framework/src/userobject/ShapeElementUserObject.C (hunk 1 above)
Comment (friedmud) [inline framework/src/userobject/ShapeElementUserObject.C]: Yep... the new stuff is getting there.

**Comment by dschwen** [comment]
> I never said it should return a Real

Well, then that is an inconsistency again. For any UO that is not inheriting from ```MooseVariableInterface``` the concept of off diagonal is moot. And again, the UOs are computing something that is conceptually different from a Jacobian that is computed by a Kernel.

Ok, I'm going to bed now. Maybe we should have a phone call over this with @permcody. I feel there was lots of productive input and we are starting to get bogged down in smaller details. I had a different idea about the scale of this change (hence the intermediate base class and the efforts to affect UOs that don't use this as little as possible). But if we can come up with a more general system that'd be great.

**Comment by friedmud** [comment]
I'm not saying not to form a list.  Re-read the comment.

I'm just saying to form the list in `subdomainChanged()` by looping over the active UserObjects and then use that list.  The Assembly doesn't need to be involved.

PLUS: I'm not even sure that your current code will work with subdomain restricted variables... since you are making one list that's used on all subdomains... some of those aren't going to exist!

Much better to just make the correct list for the current subdomain.

**Comment by dschwen** [comment]
Ha, you comment so fast half of my replies are obsolete by the time I hit the "comment" button.

**Comment by dschwen** [comment]
Ok, I like the suggestion of doing a set union on sets that live in the UOs. That is great and avoids some data duplication. And, yes, I'm aware that I have not yet treated subdomain restricted variables. Baby steps...

**Comment by friedmud** [comment]
@dschwen Everything that goes into MOOSE needs to be general and well designed.  One-off additions drain our resources in upkeep, user support and documentation.

It takes time to do this stuff right: which is why we need to discuss it _before_ writing code.

From the beginning the title of the ticket should have been `Need Jacobian Computation In UserObjects`... and the discussion should have been centered around how to achieve that in a way that fits with the rest of MOOSE, is maintainable, extensible and can benefit the most users...

**Comment by friedmud** [comment]
I would like to say: I am happy to see this going in.  I've thought a lot about it over the years... but haven't made an attempt at it.  Even though it is taking us a bit to get it right... it will be a very useful addition that many people will use.

**Comment by friedmud** [comment]
One last thing: about Postprocessors.  I _would_ like a unified interface for Postprocessor Jacobians.  Like you mention, they _should_ be Reals... so we should be able to store and process them completely internal to the library.

One note on this: Postprocessor Jacobians bring us scarily close to being able to do things like Goal Oriented Error Estimation and Mesh Adaptivity.  @roystgnr has built a lot of capability into libMesh over the years for just that... and it would be cool to be able to take advantage of it...

**Comment by friedmud** [comment]
I woke up with certainty that adding UserObject Jacobians is both a good idea... and even more complicated than I thought.

A few things:

1.  I _DO_ believe we can come up with a unified storage.  Yes, it will need to be templated (it will look a bit like the storage of Material Properties) - but it is doable.  Hell; you're the guy that added the interface to codify Material Property derivatives!
  In fact: I don't think we can merge this without it.  I don't want an interface in MOOSE where users have to think about how to set aside storage and index into it based on dof ID (something they shouldn't even know exists).  One issue is that it will need to be flexible.  UserObjects may have sparse Jacobians... or they may be fully dense.  That's the kind of stuff our users shouldn't need to think about.

2.  I don't think that just having a `computeJacobian` function is enough.  We may need to replicate all of the setup and finalization routines.  It may be ok to use `if(currently_computing_jacobian)` in those, I don't know.
  What I'm thinking about is a simple case of ElementAverageValue.  The Jacobians need to be cleared before the start (like in `initialize`).  They need to be filled in during the element loop.  BUT: they also need to be _modified_ during finalize!!  In the end you divide by the total area... and that has to be done to each Jacobian entry as well!  What does THAT look like?

**Comment by permcody** [comment]
Perhaps it's time for a phone call. Derek are you available sometime today
or tomorrow to call in and chat with the team and Daniel about the best
path forward?

**Comment by dschwen** [comment]
Ok, this adds ```ShapeElementUserObject``` lists to ```UserObjectWarehouse```. It also adds a 
```
requestJacobian(const std::string & var_name, unsigned int comp = 0);
```
method to ```ShapeElementUserObject```, which derived classes call if they need shape function prepared for a given variable. This causes ```executeJacobian(jvar)``` to be called on the class (currently executeJacobian can be called for ```jvar```s that the class has not explicitly requested. I expect the user to check the value of```jvar``` in that method!).

**Comment by friedmud** [comment]
Looking better!

Since you're pulling out the UserObjects that are `ShapeElementUserObjects` anyway... you might as well make a `std::vector` of those... and only iterate over _those_ when doing the Jacobian computation.

**Comment by friedmud** [comment]
Indeed - this is looking _much_ better!

**Comment by dschwen** [comment]
> Since you're pulling out the UserObjects that are ShapeElementUserObjects anyway... you might as well make a std::vector of those... and only iterate over those when doing the Jacobian computation.

Wait.. I thought that's what I'm doing...

**Comment by friedmud** [comment]
You're right... I see it now.  You're getting only the ShapeElementUserObjects from the warehouse to iterate over... makes sense :-)

**Comment by dschwen** [comment]
I'm thinking I could overload ```coupled``` to do the job that ```requestJacobian``` does. It takes the same parameters, users will have to call ```coupled``` anyways to compare against the jvar parameter of ```executeJacobian```, and there is no reason for a UO/PP to call coupled otherwise.

**Comment by friedmud** [comment]
I was thinking the same thing myself. Make sure to document `coupled()`
well so it's obvious what's going on.

**Comment by dschwen** [comment]
Will do. I will also let MOOSE loop over the ```_j``` and provide the global DOF index for each ```_j``` as ```_j_global```. Or is that too restrictive?

**Comment by friedmud** [comment]
I would rather _not_ pass `_j_global`.  I really hope that it's only temporary that we even have to deal with global dof indices directly for UO Jacobians... I'd rather not engrain it into the interface.

Just put a function in your base class for retrieving the global dof index based on `_j`

**Comment by friedmud** [comment]
The new commit looks good.

Next: can you put a base class parameter in that allows us to turn off Jacobian calculation for these objects from the input file?  There may be times when we don't care to compute the Jacobian (or for testing purposes).  Mark that parameter as being in the "Advanced" group...

**Comment by dschwen** [comment]
Ok. I think the most efficient way to add function to that parameter would be to query it in the UserObjectWarehouse and just suppress adding ShapeElementUserObjects with that parameter set to false to the lists.

**Comment by dschwen** [comment]
> Makes me wonder if we should embed a reference to _currently_computing_jacobian into the base class. It would be more MOOSEish. What do you think?

Yes!

**Comment by friedmud** [comment]
> Ok. I think the most efficient way to add function to that parameter would be to query it in the UserObjectWarehouse and just suppress adding ShapeElementUserObjects with that parameter set to false to the lists.

I concur.

**Comment by dschwen** [comment]
The advantage of a `_j_global` member is that I can easily set it for the user. A member function to get the global DOF id would need to get the variable passed in. I'll put it in the next commit to discuss.

**Comment by dschwen** [comment]
Hm, I could also make the advanced parameter suppress the population of the local ```_jacobian_moose_variables``` and not add the user object in the warehouse whenever ```jacobianMooseVariables()``` returns an empty set. That way I don't need an extra public method to query the status of the parameter. And we get a slightly leaner list in the warehouse if the UO never called `coupled`

**Comment by friedmud** [comment]
Yes.  Both of those.

Eventually we're going to need to be able to automatically suppress Jacobian computation by whether or not an object that is coupling to this UO cares about the Jacobian calculation.  The best way to do that is to clear out `_jacobian_moose_variables`.

**Comment by friedmud** [comment]
BTW: that last part is for _future_ work... I'm just looking off into the distance here...

File: framework/src/userobject/UserObjectWarehouse.C [hunk 5]
```diff
@@ -260,6 +279,20 @@ UserObjectWarehouse::addUserObject(MooseSharedPointer<UserObject> & user_object)
...
     }
+
+    // maintain subset consisting of the derived ShapeElementUserObjects
+    if (dynamic_cast<ShapeElementUserObject*>(raw_ptr))
+    {
+      // Extract the BlockIDs (see BlockRestrictable)
+      ShapeElementUserObject * shape_element_uo = dynamic_cast<ShapeElementUserObject *>(raw_ptr);
+
+      // Add to the list of all SideUserObjects
+      _all_shape_element_user_objects.push_back(shape_element_uo);
```
Comment (dschwen) [inline framework/src/userobject/UserObjectWarehouse.C]: I think I end up never using ```_all_shape_element_user_objects```...

**Comment by dschwen** [comment]
This should probably squashed before being merged. I'm just waiting for a final (or maybe not final) review.

File: framework/src/userobject/ShapeElementUserObject.C [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+    _jacobian_moose_variables.insert(var);
+
+  // return the variable number
+  return ElementUserObject::coupled(var_name, comp);
+}
+
+void
+ShapeElementUserObject::executeJacobianWrapper(unsigned int jvar, const std::vector<dof_id_type> & dof_indices)
+{
+  for (_j = 0; _j < _phi.size(); ++_j)
```
Comment (friedmud) [inline framework/src/userobject/ShapeElementUserObject.C]: I'm not super excited about this "Wrapper" thing.

To me the external call should be to `computeJacobian()`... and then it should call `computeQpJacobian()` like normal... (or "execute" if you _must_ continue to fight on that issue :-)

Comment (friedmud) [inline framework/src/userobject/ShapeElementUserObject.C]: Meh: Whatever... I give up at this point.  This is _fine_ for now.

We'll redo it later when we make it all more general...

**Comment by friedmud** [comment]
I think it looks fine.  I won't hit merge yet though.  @permcody @jwpeterson can one of you guys give this a brief look over and report back?  I'd just like a second opinion.

Someone else can merge this at that point... I will later today when I get out of class...

File: framework/src/userobject/ShapeElementUserObject.C (hunk 6 above)
Comment (dschwen) [inline framework/src/userobject/ShapeElementUserObject.C]: Haha! Ok, well, I tried to emulate the structure of ``computeJacobianBlock```. In there we also have a _wrapper_ method to set the local ``_j``` variable. I'm not married to that construct but yeah, why don't we leave it as is and see if we can come up with a better idea along the way.

File: framework/src/base/ComputeUserObjectsThread.C [hunk 7]
```diff
@@ -70,6 +71,32 @@ ComputeUserObjectsThread::subdomainChanged()
...
+
+    // Block Restricted ElementUserObjects
+    for (std::vector<ShapeElementUserObject *>::const_iterator it = block.begin(); it != block.end(); ++it)
+    {
+      const std::set<MooseVariable *> & mv_deps = (*it)->jacobianMooseVariables();
+      jacobian_moose_vars.insert(mv_deps.begin(), mv_deps.end());
+    }
+
+    _jacobian_moose_vars.clear();
+    std::copy(jacobian_moose_vars.begin(), jacobian_moose_vars.end(), std::back_inserter(_jacobian_moose_vars));
```
Comment (jwpeterson) [inline framework/src/base/ComputeUserObjectsThread.C]:
```
_jacobian_moose_vars.assign(jacobian_moose_vars.begin(), jacobian_moose_vars.end())
```
Would also work if you want a 1-liner.  It can either increase or decrease the size of the vector if necessary.

Comment (dschwen) [inline framework/src/base/ComputeUserObjectsThread.C]: Yeah, that is more elegant. I'll make that change once you are done reviewing.

File: test/include/userobjects/ShapeTestUserObject.h [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+class ShapeTestUserObject;
+
+template<>
+InputParameters validParams<ShapeTestUserObject>();
+
+/**
+ * Test and proof of concept class for computing UserObject Jacobians using the
+ * ShapeElementUserObject base class. This object computes the integral
+ * \f$ \int_\Omega u^2v dr \f$
+ * and builds a vector of all derivatives of the integral w.r.t. the DOFs of u and v.
```
Comment (jwpeterson) [inline test/include/userobjects/ShapeTestUserObject.h]: Did you also add a test of something that depends on `grad(u)`?  Or is that not supported yet?

File: test/include/userobjects/ShapeTestUserObject.h [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+  Real getIntegral() const { return _integral; }
+  const std::vector<Real> & getJacobian() const { return _jacobian_storage; }
+  ///@}
+
+protected:
+  Real _integral;
+  std::vector<Real> _jacobian_storage;
+
+  /// Bit mask for testing purposes (bits get set if executeJacobian is called for a variable)
+  unsigned int _execute_mask;
```
Comment (jwpeterson) [inline test/include/userobjects/ShapeTestUserObject.h]: Do we want users to copy/paste code like this `_execute_mask` business?  I know it is for a test but they will probably try to emulate it in their own user objects...

File: test/include/userobjects/ShapeTestUserObject.h (hunk 8 above)
Comment (dschwen) [inline test/include/userobjects/ShapeTestUserObject.h]: That is reported, but I'm not explicitly testing against that. I was assuming that if _phi gets initialized correctly the same should be true for _grad_phi...

Comment (dschwen) [inline test/include/userobjects/ShapeTestUserObject.h]: That being said I could definitely add this. I should probably also write a Jacobian test.

**Comment by jwpeterson** [comment]
Looks fine from a coding conventions standpoint.  I am not as familiar with the overall design as you guys so I won't comment on that...

File: test/include/userobjects/ShapeTestUserObject.h (hunk 9 above)
Comment (dschwen) [inline test/include/userobjects/ShapeTestUserObject.h]: No we don't. One option would be to split this object into two test classes. One that actually computes a Jacobian, and one that only checks if execution happens.

**Comment by friedmud** [comment]
BTW: I vote for NOT squashing this.  The individual commits show the progression here... and the comments on each commit will be useful when we try to go further for future work...

**Comment by permcody** [comment]
:+1:

**Comment by dschwen** [comment]
Ok, I split the test object into two and added a Kernel that actually _uses_ the Jacobian data to build the Jacobian matrix, and whaddaya know, it's **wrong**.
I've pushed this as I may need a second pair of eyes on this...

**Comment by jwpeterson** [comment]
> it's wrong.

I guess it's a good thing we tested it.

**Comment by friedmud** [comment]
Hmmm.... I can't see why it's wrong.  The math appears correct.

I'm going to sleep on it and see what I come up with in the morning....

**Comment by dschwen** [comment]
Yeah, the math looks right. I think it is a bookkeeping issue. If I change the `v` Jacobian terms in `ExampleShapeElementUserObject` to something purposely wrong the Jacobian tester tells me that the `u` Jacobian gets worse! 
Something in the way I wrangle thos DOF IDs must be messed up.

**Comment by dschwen** [comment]
I think the DOF accounting is correct after all. There must be something wrong with the math conceptually.

If I change the test to 1x1 element mesh the Jacobian is correct.

**Comment by permcody** [comment]
@dschwen - I'm closing this branch since it can't be merged as is. Feel free to revive and reopen when you get back to this work.

**Comment by dschwen** [comment]
Yeah, this will happen soon, and I may need some help with it. I rebased it onto current devel, which worked fine.
